raven/processes/wps_shape_selection.py

```diff
@@ -0,0 +1,144 @@
+            ComplexOutput('upstream_basins', 'IDs for all immediate upstream basins',
+                          abstract='List of all first-level tributary basins by their HydroBASINS ID',
+                          supported_formats=[FORMATS.JSON])
+        ]
+
+        super(ShapeSelectionProcess, self).__init__(
+            self._handler,
+            identifier="shape-selection",
+            title="Shape Selection",
+            version="1.0",
+            abstract="Return a feature within a polygonal vector file based on lat/lon coordinates.",
+            metadata=[],
+            inputs=inputs,
+            outputs=outputs,
+            status_supported=True,
+            store_supported=True)
+
+    def _handler(self, request, response):
+
+        # collect_upstream = request['collect_upstream']
+        # lonlat = request['lonlat_coordinate']
+        # crs = request['crs']
+        # shape_url = request['shape']
+
+        collect_upstream = request.inputs['collect_upstream'][0].data
+        lonlat = request.inputs['lonlat_coordinate'][0].data
+        crs = request.inputs['crs'][0].data
+        shape_url = request.inputs['shape'][0].file
+
+        try:
+            lon, lat = tuple(map(float, re.findall(r'[-+]?[0-9]*\.?[0-9]+', lonlat)))
+        except Exception as e:
+            msg = 'Failed to parse geo-coordinates {0}: {1}'.format(lonlat, e)
+            LOGGER.error(msg)
+            return response
+
+        types = ['.tar', '.zip', '.7z']
+        extensions = ['.gml', '.shp', '.geojson', '.json']  # '.gpkg' requires more handling
+        shape_url = archive_sniffer(shape_url, working_dir=self.workdir, archive_types=types, extensions=extensions)
+
+        # pfaf = ''
+
+        basin = []
+        upstream_basins = []
+        properties = []
+        location = Point(lon, lat)
+
+        try:
+            with fio.open(shape_url, 'r', crs=from_epsg(crs)) as src:
+                for feature in src:
+                    geom = shape(feature['geometry'])
+
+                    if geom.contains(location):
+                        # pfaf = feature['properties']['PFAF_ID']
+
+                        basin = [feature['properties']['HYBAS_ID']]
+                        prop = {'id': feature['id']}
+                        prop.update(feature['properties'])
+                        prop.update(feature['geometry'])
+                        properties.append(prop)
+
+                    else:
+                        continue
+
+                if collect_upstream:
+                    # 'PFAF_ID' can also technically be used to described the drainage network. See HydroBASINS docs.
+                    # pfaf_start, pfaf_end = str(pfaf)[0:3], str(pfaf)[3:]
+
+                    upstream = filter(lambda f: feature['properties']['NEXT_DOWN'] == basin[0], src)
+                    for up in iter(upstream):
+                        upstream_basins.append(up['properties']['HYBAS_ID'])
```

Comment:
Yes. That's the idea behind this function. The reason why I haven't included more levels is because I wasn't certain that that's what we wanted and because I wanted to evade creating a potential logic bomb that could continuously loop to return 1/4 of all listed features.

I was reading the other day that if we leverage geopandas, we might be able to perform SQL-like selection and not have to constantly re-iterate through the file.

One thing I'd like to know is if the desired end result is a union of all collected features. If so, this would drastically lower the size of the geojson output and make it a bit easier to collect.

Reply:
Not clear to me either, but my guess is that we need to go up all the way. 
Yes, union of all features is good.

Reply:
This has been split into the `hydrobasins_shape_selection` function and is much better handled now.